src/sql/src/session/vars.rs

```diff
@@ -270,14 +271,26 @@ impl SessionVar {
         }
     }
 
-    pub fn end_transaction(&mut self, action: EndTransactionAction) {
-        self.local_value = None;
+    /// Returns a possibly new SessionVar if this needs to mutate at transaction end.
+    #[must_use]
+    pub fn end_transaction(&self, action: EndTransactionAction) -> Option<Self> {
+        if !self.is_mutating() {
+            return None;
+        }
+        let mut next: Self = self.clone();
+        next.local_value = None;
         match action {
-            EndTransactionAction::Commit if self.staged_value.is_some() => {
-                self.session_value = self.staged_value.take()
+            EndTransactionAction::Commit if next.staged_value.is_some() => {
+                next.session_value = next.staged_value.take()
             }
-            _ => self.staged_value = None,
+            _ => next.staged_value = None,
         }
+        Some(next)
+    }
+
+    /// Whether this Var needs to mutate at the end of a transaction.
+    pub fn is_mutating(&self) -> bool {
```

Comment:
what do you think about calling this `is_dirty(...)`?

Reply:
Eh, I don't really like either? I'm going to merge as is and if someone feels strongly they can go through another CI cycle.

Reply:
No worries, `is_mutating(...)` works!